src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/jose/ValidateJwtProof.kt

```diff
@@ -37,43 +37,52 @@ import com.nimbusds.jwt.proc.DefaultJWTProcessor
 import com.nimbusds.jwt.proc.JWTProcessor
 import eu.europa.ec.eudi.pidissuer.domain.*
 import eu.europa.ec.eudi.pidissuer.port.input.IssueCredentialError
+import eu.europa.ec.eudi.pidissuer.port.out.jose.DecryptCNonce
 import java.security.interfaces.ECPublicKey
 import java.security.interfaces.EdECPublicKey
 import java.security.interfaces.RSAPublicKey
 import kotlin.time.Duration.Companion.seconds
 import kotlin.time.DurationUnit
 
-context (Raise<IssueCredentialError.InvalidProof>)
-fun validateJwtProof(
-    credentialIssuerId: CredentialIssuerId,
-    unvalidatedProof: UnvalidatedProof.Jwt,
-    expectedCNonce: CNonce,
-    credentialConfiguration: CredentialConfiguration,
-): CredentialKey {
-    val proofType = credentialConfiguration.proofTypesSupported[ProofTypeEnum.JWT]
-    ensureNotNull(proofType) {
-        IssueCredentialError.InvalidProof("credential configuration '${credentialConfiguration.id.value}' doesn't support 'jwt' proofs")
+/**
+ * Validator for JWT Proofs.
+ */
+internal class ValidateJwtProof(
+    private val credentialIssuerId: CredentialIssuerId,
+    private val decryptCNonce: DecryptCNonce,
+) {
+    context (Raise<IssueCredentialError.InvalidProof>)
+    suspend operator fun invoke(
+        unvalidatedProof: UnvalidatedProof.Jwt,
+        credentialConfiguration: CredentialConfiguration,
+    ): Pair<CredentialKey, CNonce> {
+        val proofType = credentialConfiguration.proofTypesSupported[ProofTypeEnum.JWT]
+        ensureNotNull(proofType) {
+            IssueCredentialError.InvalidProof("credential configuration '${credentialConfiguration.id.value}' doesn't support 'jwt' proofs")
+        }
+        check(proofType is ProofType.Jwt)
+
+        return validate(unvalidatedProof, proofType)
     }
-    check(proofType is ProofType.Jwt)
 
-    return validateJwtProof(credentialIssuerId, unvalidatedProof, expectedCNonce, proofType)
+    context (Raise<IssueCredentialError.InvalidProof>)
+    private suspend fun validate(
```

Comment:
Let's extract a new method that returns the pair of `CredentialKey` and the value of the `nonce` claim, if present.

```kotlin

private suspend fun Raise<IssueCredentialError.InvalidProof>.credentialKeyNonce(
    unvalidatedProof: UnvalidatedProof.Jwt,
    proofType: ProofType.Jwt,
): Pair<CredentialKey, String?> = TODO()
```

Then we can refactor validate() to enforce (or not) the presence of a valid `c_nonce`

To do so, we need to pass an optional decrypt

```kotlin

private suspend fun Raise<IssueCredentialError.InvalidProof>.validate(
    unvalidatedProof: UnvalidatedProof.Jwt,
    proofType: ProofType.Jwt,
    decryptCNonce: DecryptCNonce?
): Pair<CredentialKey, CNonce?> {
    
    
    val (key, nonce) = credentialKeyNonce(unvalidatedProof, proofType)
    if (decryptCNonce != null) {
        // require nonce not null and valie
        }
    
}

```